### PR TITLE
Clm/prompts - authored by Tom in an amazing fit of coding frenzy!! 

### PR DESCRIPTION
**Pull request by cleemullins**
Initial cut at a text recognizer. 

This recognizer is loosely based on the Node Recognizer that Steve has in the [botbuilder-toybox](https://github.com/Stevenic/botbuilder-toybox/blob/master/packages/botbuilder-toybox-prompts/src/textPrompt.ts
) repo. 

Adjustments have been made to make it more C# friendly. 

Comments welcome.

### CODE REVIEW

**Comment by garypretty** [comment]
@brandonh-msft Really like the direction you are taking here and I agree that wrapping the prompt and response for the user would make a lot of sense. Great stuff.  @cleemullins I will try and get some time to review your initial work today / tomorrow and feedback :)

**Comment by Stevenic** [comment]
As @cleemullins said this is inspired by a light weight prompt system I built for the JS side of things.  This is like my 12th attempt at prompts for v4 and if you look through my code samples I think you'll see that it's a pretty well balanced system. Chris implemented the `TextPrompt` but to be honest this is by far the simplest prompt so it's kind of difficult to see what the fuss is all about. You need to look at something like the `ChoicePrompt` or the `DatetimePrompt` for things to really click.

When you define a prompt using this system you're always going to get back an object with at least a `Prompt()` and `Recognize()` method. The signature for Prompt() may very because formatting a choice prompt is radically different then formatting a text prompt but you as the developer know that anytime you call Prompt() it's going to send a properly formatted "prompt" to the user. This is significant because  this is not always the same as sending a "message" to the user. On the Cortana channel for instance the prompt will automatically set the `inputHint` to `expectingInput` so it's not always the same as calling `reply()`.  The other big thing is that calling the prompts `Prompt()` method is completely optional and if you need to do something radical and want to send your prompt using `reply()` you can.

On the recognition/validation side of things the prompts base line `Recognize()` method will always use the most recent reply to attempt to recognize what ever the prompt is designed to recognize. On the JS side you get the parsed value back if its recognized or `undefined` if it's not. I see that Chris is returning a flag back to indicate the value was recognized which I guess is ok if that's more what's expected in C#. The thing is the default result from `Recognize()` is a recognized value but it hasn't been validated.  To add validation you can pass in a custom validator when you create the prompt and this validator will get called every time Recognize() is called. Your validator gets passed the context object and recognized value which again may be undefined if it wasn't recognized. This gives your code an opportunity to not only add additional constraints to the input but you can also change the value returned by the `Recognize()` method to something else.  For instance, the DateTime prompts recognizer by default returns the raw LUIS TIMEX's that were recognized. Adding a custom validator you can both ensure that you only allow dates in the future and then change the value returned to be a simple DateTime.

The system feels a lot more balanced then our previous attempts at prompts for a couple of reasons:

- Its one notch above working with the raw primitives directly. You get the benefits of an abstraction but you don't feel like your handcuffed by the abstraction thanks to the super flexible validator pattern.
- There's not built-in stickyness. It's tempting to want to just provide a turn-key prompt the way v3 does but this is really the point at which things start feeling like there's magic going on and your hands are tied in certain situation.
- The prompts themselves don't know how to re-prompt. A lot of our previous attempts tried to build in the re-prompt logic and again that's perceived as magic and there are just too many possible ways you want to reprompt the user that there's no one-size-fits all solution.

**Comment by cleemullins** [comment]
@brandonh-msft How did you like the Tuple coming back?
`var (Passed, Value) = await context.ValidatePromptReply(MinLengthValidator);`

While I thought the resulting code was very readable, I'm worried that I see few people using Tuples. The other solutions all had more weird stuff like out parameters or excessive null checks. 

I'll check out your thoughts (in your branch) tonight.

**Comment by brandonh-msft** [comment]
I'm not opposed to Tuples at all, however it's worth noting they'll result in a breaking change should we add/remove a value to them. Perhaps a `PromptResult<T>` would be more appropriate, having a `Passed` and `T Value` on it. Then if we choose to add something else (ie: maybe store any exception in an `Exception Error` property) , it's not breaking folks.

**Comment by Stevenic** [comment]
What do you normally get back from a `TryParse()`? That's actually a more equivalent output here. Calling `Recognize()` you're essentially attempting to parse the users input and that either succeeds or fails because  the input wasn't properly formatted.

**Comment by brandonh-msft** [comment]
Kinda where my head initially went, as well @Stevenic - I'd be fine just getting back the value w/ an Exception it failed, or as you pointed out, treating it like `TryParse` where you get back a boolean with an `out` parameter w/ the value (should the bool resolve to `true`)

the added advantage I see to the return object route or the exception route is you can provide the developer the **reason** for the failure.

**Comment by cleemullins** [comment]
TryParse returns a boolean for true/false with an OUT parameter for the actual value. This behavior has been there since .Net 1.0, and likley has this format due to lack of Tuples.

I'm ok with either of:
1. Tuples
2. PromptResult<T>

The PromptResult<T> is certainly more future proof, as it's quite extensible; adding reasons ("errorReasonText", "errorReasonCard", "errorReasonSpeak",  etc) , multiple languages (errorText[en-us] = "Number must be between 1 and 10", errorText[zn] = "该数字必须在1到10之间") custom handlers, and the like all seems very practical.  Code-readability wise, the Tuples visually seemed to read cleaner to me, but it's close.

I don't like the Exception coming back, as the "Validation Failed" is likely to happen more than "Validation Passed" and abusing Exceptions for flow control is generally considered an anti-pattern.

File: libraries/Microsoft.Bot.Builder.Prompts/Microsoft.Bot.Builder.Prompts.csproj [hunk 1]
```diff
@@ -4,6 +4,13 @@
     <TargetFramework>netstandard2.0</TargetFramework>
   </PropertyGroup>
 
+  <ItemGroup>
+    <PackageReference Include="Microsoft.Recognizers.Text.DateTime" Version="1.0.0.36" />
+    <PackageReference Include="Microsoft.Recognizers.Text.Number" Version="1.0.0.36" />
+    <PackageReference Include="Microsoft.Recognizers.Text.NumberWithUnit" Version="1.0.0.36" />
+    <PackageReference Include="Microsoft.Recognizers.Text.Sequence" Version="1.0.0.36" />
+  </ItemGroup>
+
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder.Prompts/Microsoft.Bot.Builder.Prompts.csproj]: Did ya notice that these packages are (as of Monday) .Net Standard 2?? The Build warnings are gone!

File: tests/Microsoft.Bot.Builder.Prompts.Tests/AgePromptTests.cs [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+namespace Microsoft.Bot.Builder.Prompts.Tests
+{
+    [TestClass]
+    [TestCategory("Prompts")]
+    [TestCategory("Age Prompts")]
+    public class AgePromptTests
+    {
+        [TestMethod]
+        public async Task AgePrompt_Test()
```
Comment (cleemullins) [inline tests/Microsoft.Bot.Builder.Prompts.Tests/AgePromptTests.cs]: This is... beautiful.

Comment (cleemullins) [reply tests/Microsoft.Bot.Builder.Prompts.Tests/AgePromptTests.cs]: All. These. Tests. 

Love it.

File: tests/Microsoft.Bot.Builder.Prompts.Tests/PercentagePromptTests.cs [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        [TestMethod]
+        public async Task PercentagePrompt_Validator()
+        {
+            TestAdapter adapter = new TestAdapter()
+                .Use(new ConversationState<TestState>(new MemoryStorage()));
+
+            await new TestFlow(adapter, async (context) =>
+            {
+                var state = ConversationState<TestState>.Get(context);
+                var numberPrompt = new PercentagePrompt(Culture.English, async (ctx, result) =>  result.Value > 10);
```
Comment (cleemullins) [inline tests/Microsoft.Bot.Builder.Prompts.Tests/PercentagePromptTests.cs]: It's really, really nice how simple this is to use. Validation and everything.

File: tests/Microsoft.Bot.Builder.Prompts.Tests/NumberPromptTests.cs [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+                    var numberPrompt = new NumberPrompt<float>(Culture.English);
+                    if (!state.InPrompt)
+                    {
+                        state.InPrompt = true;
+                        await numberPrompt.Prompt(context, "Gimme:");
+                    }
+                    else
+                    {
+                        var result = await numberPrompt.Recognize(context);
+                        if (result == null)
```
Comment (cleemullins) [inline tests/Microsoft.Bot.Builder.Prompts.Tests/NumberPromptTests.cs]: The only thing I don't like about these prompts is that Null Checking is required to use them. Is there a clear way to get away from that? The Tuple approach I had seemed much clearer, and the RecognizeResult.DidPass approach that Brandon suggested also seemed valid.

**Comment by cleemullins** [comment]
Apparently I cannot approve this PR, as I was the original requestor. 

Nonetheless, I approve. Code looks fantastic. Merge it!

File: tests/Microsoft.Bot.Builder.Prompts.Tests/NumberPromptTests.cs (hunk 4 above)
Comment (brandonh-msft) [reply tests/Microsoft.Bot.Builder.Prompts.Tests/NumberPromptTests.cs]: again, `PromptResult<T>` would solve this.

**Comment by cleemullins** [comment]
Yes. I think I called that out. :)

________________________________

**Pull request merged**